File: 2016/Day01/day01.pl
#!/usr/bin/env perl

# Advent of Code 2016: Day 1
# http://adventofcode.com/2016/day/1

use strict;
use warnings;

# Part 1
my $heading = 0;
my @blocks = qw/0 0 0 0/; # North East South West

# Part 2
my %city = ( 0 => { 0 => 1 } );
my ($x, $y) = (0, 0);
my $foundHQ = 0;

open my $fh, '<', './input' or die "$0: can't open './input' for reading: $!";
foreach my $step (split(/, /, do { local $/; <$fh>; }))
{
  my $count = substr($step, 1);
  if (substr($step, 0, 1) eq 'R')
  {
    $heading = ++$heading % 4
  }
  else
  {
    $heading = --$heading % 4
  }
  $blocks[$heading] += $count;

  unless ($foundHQ) {
    foreach (1..$count)
    {
      ++$y if $heading == 0;
      ++$x if $heading == 1;
      --$y if $heading == 2;
      --$x if $heading == 3;

      if (++$city{$x}{$y} == 2)
      {
        $foundHQ = 1;
        last;
      }
    }
  }
}

my $distance = abs($blocks[0] - $blocks[2]) + abs($blocks[1] - $blocks[3]);
my $distance2 = abs($x + $y);
print "Part 1: $distance\n";
print "Part 2: $distance2 ($x,$y)\n";

File: 2016/Day01/day01.cpp
/*
 * Advent of Code 2016: Day 1
 * http://adventofcode.com/2016/day/1
 */

#include <iostream>
#include <fstream>
#include <cmath>
#include <unordered_map>
#include <string>

using namespace std;

enum { NORTH, EAST, SOUTH, WEST };

int main(int argc, char *argv[])
{
  // Part 1
  int count, heading = NORTH, x = 0, y = 0;
  char rot;

  // Part 2
  unordered_map<int, unordered_map<int, int>> city = {
    {0, {
      {0, 1}
    }}
  };
  int distanceHQ = 0;
  string hq;

  ifstream input("./input");
  while (input.get(rot))
  {
    if (rot == 'R')
    {
      if (++heading == 4) heading = NORTH;
    }
    else
    {
      if (--heading == -1) heading = WEST;
    }

    input >> count;
    for (int i = 0; i < count; ++i)
    {
      switch (heading)
      {
        case NORTH: ++y; break;
        case EAST:  ++x; break;
        case SOUTH: --y; break;
        case WEST:  --x; break;
      }

      if (!distanceHQ)
      {
        if (++city[x][y] == 2)
        {
          distanceHQ = abs(x + y);
          hq = to_string(x) + ", " + to_string(y);
        }
      }
    }

    input.ignore(2, ' ');
  }

  cout << "Part 1: " << abs(x - y) << endl;
  cout << "Part 2: " << distanceHQ << " (" << hq << ")\n";

  return 0;
}
